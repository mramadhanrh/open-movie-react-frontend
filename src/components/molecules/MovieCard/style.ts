import styled from 'styled-components';

export const Card = styled.button`
  border: none;
  background: #ffffff;
  padding: 0;
  margin: 10px 0;
  width: 100%;
  min-height: 20px;
  box-shadow: 0px 2px 20px rgba(0, 0, 0, 0.16);
  border-radius: 10px;
  max-width: 300px;
  text-align: left;
  overflow: hidden;
`;

export const CardImage = styled.img<{ src?: string }>`
  width: 100%;
`;

export const InfoWrapper = styled.div`
  padding: 10px 10px 15px;

  & > h1 {
    font-size: 24px;
    margin: 0;
  }
`;

export const BadgesWrapper = styled.div`
  display: flex;
  text-transform: capitalize;
  & > div {
    margin-right: 5px;
  }
`;
