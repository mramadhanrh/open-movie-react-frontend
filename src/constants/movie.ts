export const MovieEmptyPicture = {
  label: 'N/A',
  src: 'https://images.unsplash.com/photo-1578328819058-b69f3a3b0f6b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=967&q=80',
};
