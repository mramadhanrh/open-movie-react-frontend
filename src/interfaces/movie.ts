export enum MovieResultType {
  Movie = 'movie',
  Series = 'series',
  Episode = 'episode',
}

export interface MovieItem {
  Title: string;
  Year: string;
  imdbID: string;
  Type: MovieResultType;
  Poster: string;
}

export interface MovieSearchResponse {
  Search: MovieItem[];
  totalResults: string;
  Response: 'True' | 'False';
  Error?: string;
}

export interface MovieRating {
  Source: string;
  Value: string;
}

export interface MovieDetailResponse {
  Title: string;
  Year: string;
  Rated: string;
  Released: string;
  Runtime: string;
  Genre: string;
  Director: string;
  Writer: string;
  Actors: string;
  Plot: string;
  Language: string;
  Country: string;
  Awards: string;
  Poster: string;
  Ratings: MovieRating[];
  Metascore: string;
  imdbRating: string;
  imdbVotes: string;
  imdbID: string;
  Type: MovieResultType;
  DVD: string;
  BoxOffice: string;
  Production: string;
  Website: string;
  Response: string;
  Error?: string;
}
