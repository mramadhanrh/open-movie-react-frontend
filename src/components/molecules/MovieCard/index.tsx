import { FC } from 'react';
import { MovieResultType } from '../../../interfaces/movie';
import Badges from '../../atoms/Badges';
import { BadgesWrapper, InfoWrapper, Card, CardImage } from './style';

interface MovieCardProps {
  title?: string;
  picture?: string;
  type?: MovieResultType;
  year?: string;
}

const MovieCard: FC<MovieCardProps> = ({ title, picture, type, year }) => (
  <Card>
    <CardImage src={picture} />
    <InfoWrapper>
      <h1>{title}</h1>
      <BadgesWrapper>
        <Badges color="primary">{type}</Badges>
        <Badges>{year}</Badges>
      </BadgesWrapper>
    </InfoWrapper>
  </Card>
);

export default MovieCard;
