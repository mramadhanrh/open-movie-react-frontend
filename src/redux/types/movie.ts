import { MovieItem } from '../../interfaces/movie';

export enum MovieType {
  SetMovieList = '@move/set-list',
}

export interface MovieSetList {
  type: MovieType.SetMovieList;
  payload: MovieItem[];
}

export type MovieAction = MovieSetList;
