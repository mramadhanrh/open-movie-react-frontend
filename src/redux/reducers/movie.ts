import { MovieItem } from '../../interfaces/movie';
import { MovieAction, MovieType } from '../types/movie';

interface MovieState {
  list: MovieItem[];
}

const initialState = {
  list: [],
};

const { SetMovieList, AddMovieList } = MovieType;

const movie = (
  state: MovieState = initialState,
  action: MovieAction,
): MovieState => {
  const { type, payload } = action;

  const setMovieList = (): MovieState => ({
    ...state,
    list: payload,
  });

  const addMovieList = (): MovieState => ({
    ...state,
    list: [...state.list, ...(payload || [])],
  });

  switch (type) {
    case SetMovieList:
      return setMovieList();
    case AddMovieList:
      return addMovieList();
    default:
      return state;
  }
};
export default movie;
