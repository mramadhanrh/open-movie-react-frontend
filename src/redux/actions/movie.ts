import { AxiosResponse } from 'axios';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  MovieItem,
  MovieResultType,
  MovieSearchResponse,
} from '../../interfaces/movie';
import { getMovieList } from '../../services/searchMovie';
import { AppStateType } from '../reducers';
import { MovieAddList, MovieSetList, MovieType } from '../types/movie';

export const setMovieList = (list: MovieItem[]): MovieSetList => ({
  type: MovieType.SetMovieList,
  payload: list,
});

export const setAddMovieList = (list: MovieItem[]): MovieAddList => ({
  type: MovieType.AddMovieList,
  payload: list,
});

export const loadMovieInitList =
  (search: string, page = 1, type?: MovieResultType) =>
  async (
    dispatch: ThunkDispatch<AppStateType, null, AnyAction>,
  ): Promise<
    [MovieSearchResponse | null, AxiosResponse<MovieSearchResponse> | null]
  > => {
    try {
      const res = await getMovieList(search, page, type);
      dispatch(setMovieList(res.data.Search));

      return [res.data, null];
    } catch (e) {
      return [null, e];
    }
  };

export const loadMovieAddList =
  (search: string, page = 1, type?: MovieResultType) =>
  async (
    dispatch: ThunkDispatch<AppStateType, null, AnyAction>,
  ): Promise<
    [MovieSearchResponse | null, AxiosResponse<MovieSearchResponse> | null]
  > => {
    try {
      const res = await getMovieList(search, page, type);
      dispatch(setAddMovieList(res.data.Search));

      return [res.data, null];
    } catch (e) {
      return [null, e];
    }
  };
