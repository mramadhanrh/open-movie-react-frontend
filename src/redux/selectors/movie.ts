import { MovieDetailResponse, MovieItem } from '../../interfaces/movie';
import { AppStateType } from '../reducers';

export const movieListSelector = (state: AppStateType): MovieItem[] =>
  state.movie.list;

export const movieDetailSelector = (
  state: AppStateType,
  id: string,
): MovieDetailResponse => state.movie.detail[id];
