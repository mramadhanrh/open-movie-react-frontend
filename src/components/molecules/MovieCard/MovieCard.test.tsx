import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MovieCard from '.';
import defaultTheme from '../../../constants/theme';
import { MovieResultType } from '../../../interfaces/movie';

describe('MovieCard', () => {
  const mockFn = jest.fn();
  const cardComp = (
    <MovieCard
      id="1"
      title="Title"
      picture="picture-url"
      type={MovieResultType.Movie}
      year="1998"
      onClick={mockFn}
    />
  );

  test('match snapshot', () => {
    expect(toJson(shallow(cardComp))).toMatchSnapshot();
  });

  test('container is clickable', () => {
    const wrapper = shallow(cardComp);
    wrapper.simulate('click');

    expect(mockFn).toHaveBeenCalled();
  });

  test('render content text', () => {
    const wrapper = mount(cardComp);

    expect(wrapper.text().includes('Title')).toBe(true);
    expect(wrapper.text().includes('1998')).toBe(true);
    expect(wrapper.text().includes(MovieResultType.Movie)).toBe(true);
    expect(wrapper.find('img').prop('src')?.includes('picture-url')).toBe(true);
  });
});
