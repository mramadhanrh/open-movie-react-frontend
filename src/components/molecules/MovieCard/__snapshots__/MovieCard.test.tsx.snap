// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard match snapshot 1`] = `
<styled.button
  onClick={[Function]}
>
  <styled.img
    src="picture-url"
  />
  <styled.div>
    <h1>
      Title
    </h1>
    <styled.div>
      <styled.div
        color="primary"
      >
        movie
      </styled.div>
      <styled.div>
        1998
      </styled.div>
    </styled.div>
  </styled.div>
</styled.button>
`;
